modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/TokensAutodetector/TransactedTokensAutodetector.swift

```diff
@@ -0,0 +1,154 @@
+        let alreadyAddedContracts = tokensDataStore.tokens(for: [session.server]).map { $0.contractAddress }
+        let deletedContracts = tokensDataStore.deletedContracts(forServer: session.server).map { $0.contractAddress }
+        let hiddenContracts = tokensDataStore.hiddenContracts(forServer: session.server).map { $0.contractAddress }
+        let delegateContracts = tokensDataStore.delegateContracts(forServer: session.server).map { $0.contractAddress }
+
+        return detectedContracts - alreadyAddedContracts - deletedContracts - hiddenContracts - delegateContracts
+    }
+}
+
+extension TransactedTokensAutodetector {
+    final class ContractInteractionsSchedulerProvider: SchedulerProvider {
+        private let session: WalletSession
+        private let apiNetworking: ApiNetworking
+        private var storage: ContractInteractionsPaginationStorage
+        private let subject = PassthroughSubject<Result<[AlphaWallet.Address], PromiseError>, Never>()
+        private let tokenType: Eip20TokenType
+
+        let interval: TimeInterval
+        var name: String = ""
+        var operation: AnyPublisher<Void, PromiseError> {
+            return fetchPublisher()
+        }
+
+        var publisher: AnyPublisher<Result<[AlphaWallet.Address], PromiseError>, Never> {
+            subject.eraseToAnyPublisher()
+        }
+
+        init(session: WalletSession,
+             apiNetworking: ApiNetworking,
+             storage: ContractInteractionsPaginationStorage,
+             tokenType: Eip20TokenType,
+             interval: TimeInterval = 0,
+             name: String = "") {
+
+            self.name = name
+            self.interval = interval
+            self.tokenType = tokenType
+            self.storage = storage
+            self.session = session
+            self.apiNetworking = apiNetworking
+        }
+
+        private func fetchPublisher() -> AnyPublisher<Void, PromiseError> {
+            let startBlock = storage
+                .transactionsPagination(server: session.server, tokenType: tokenType)
+
+            return fetchOperation(walletAddress: session.account.address, startBlock: startBlock)
+                .handleEvents(receiveOutput: { [weak self] response in
+                    self?.handle(response: response)
+                }, receiveCompletion: { [weak self] result in
+                    guard case .failure(let e) = result else { return }
+                    self?.handle(error: e)
+                }).mapToVoid()
+                .eraseToAnyPublisher()
+        }
+
+        private func fetchOperation(walletAddress: AlphaWallet.Address,
+                                    startBlock: Int?) -> AnyPublisher<UniqueNonEmptyContracts, PromiseError> {
+            switch tokenType {
+            case .erc1155:
+                return apiNetworking.erc1155TokenInteractions(walletAddress: walletAddress, startBlock: startBlock)
```

Comment:
In `CovalentApiNetworking`, `erc1155TokenInteractions(walletAddress:startBlock:)` returns `.empty()`, but does `EtherscanCompatibleApiNetworking`'s implementation handle when not every chain supports it?

Also, does every chain that is `EtherscanCompatibleType.etherscan` actually support `action=token1155tx`?

How does the code know to fallback to events? Or do we always do both? If we do that, can we change it so that we prefer getting them from blockchain explorer APIs (i.e. Etherscan and friends) and if that isn't available, then we fallback to events? We could change this in a separate PR, since we already do both for `ERC20` and `ERC721` too right? But by doing both for ERC-1155, does it confuse our token ID-count bookkeeping?

Reply:
> How does the code know to fallback to events?

it works without fallback to events.

Reply:
> In CovalentApiNetworking, erc1155TokenInteractions(walletAddress:startBlock:) returns .empty(), but does EtherscanCompatibleApiNetworking's implementation handle when not every chain supports it?

Covalent, as well as Oklink they don't have separate endpoint for retrieving such transactions, might be need to update with additionaly filter logic for each one

Reply:
> Also, does every chain that is EtherscanCompatibleType.etherscan actually support action=token1155tx?

haven't check for all of available in the app, but some of them work only when startBlock, endBlock provided, also need to handle this. I don't know is that acceptable if we would take end block as some number (100 for example to make that range, also with checking if api accepts such rangle)

Reply:
I'm guessing that not every chain on Etherscan will support it. If that's the case, we'll have to hardcode the supported chains.

> but some of them work only when startBlock, endBlock provided, also need to handle this. I don't know is that acceptable if we would take end block as some number (100 for example to make that range, also with checking if api accepts such rangle)

Ouch. Yes I guess we'll have to paginate and figure out if the limits are the same for each chain